File: Linked List/19. Remove Nth Node From End of List/19. Remove Nth Node From End of List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def removeNthFromEnd(self, head, n):
        """
        :type head: ListNode
        :type n: int
        :rtype: ListNode
        """

        fast=head    # 快指针
        slow=head    # 慢指针

        # 1. 快指针移动n步
        for i in range(n):
            fast=fast.next
        
        # 2. 判断快指针是否移动到末尾，即fast==None，若fast==None，则待删除的是头结点，特殊处理删去头结点
        if fast==None:
            head=head.next    # 与C或者C++不同，python无需手动释放内存
            return head
        # 2. 若fast!=None，则同时移动快指针和慢指针，直到fast->next==None，这时slow指向待删除结点的前驱结点
        else:
            while fast.next!=None:
                fast=fast.next
                slow=slow.next
            slow.next=slow.next.next
            return head
            
        

File: Linked List/19. Remove Nth Node From End of List/19. Remove Nth Node From End of List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

// 非常基础的链表问题，同样使用双指针法进行求解: 
// (1) 创建一个快指针fast和慢指针slow指向链表头结点head
// (2) 将快指针向前移动k步
// (3) 同时循环移动快指针和慢指针一步，直到快指针指向NULL，这时慢指针恰好指向链表的倒数第k个结点，删除慢指针这时所指向的结点即可

class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode *fast=head;      // 快指针
        ListNode *slow=head;      // 慢指针
        ListNode *slowprev=NULL;  // 慢指针的前驱结点，因为要删除慢指针所指向的结点，因此需要一个额外的指针指向慢指针的前驱结点

        for(int i=0;i<n;i++)
        {
            fast=fast->next;
        }

        while(fast)
        {
            slowprev=slow;
            fast=fast->next;
            slow=slow->next;
        }

        if(slowprev)
        slowprev->next=slow->next;
        else
        head=head->next;
        
        delete slow;  
        
        return head;
    }
};
